### PR TITLE
De-duplicate Linux builds with priority of MUSL targets

### PR DESCRIPTION
**Pull request by T-256**
## Summary
Ref https://github.com/astral-sh/uv/issues/4122#issuecomment-2154280414 musllinux targets are superset of gnu targets.

Should we document it that why we use musl builds?

## Test Plan
https://github.com/T-256/uv/actions/runs/9416773652/job/25940621883

### CODE REVIEW

**Comment by ofek** [comment]
I think it would be nice to document!

File: README.md [hunk 1]
```diff
@@ -506,6 +506,10 @@ uv has Tier 2 support (["guaranteed to build"](https://doc.rust-lang.org/beta/ru
 - Linux (i686)
 - Linux (s390x)
 
+The ABI used for Linux platform is MUSL, except for _PPC64_ and _s390x_ archs used GNU ABI.
```
Comment (henryiii) [inline README.md:509]:
```suggestion
The ABI used for Linux platform is MUSL with statically linked C-Runtime, except for _PPC64_ and _s390x_ archs which are only available with dynamically linked GNU ABI.
```

**Comment by henryiii** [comment]
I think the wheels also need to be dual tagged. You can tell maturin to do this.

**Comment by T-256** [comment]
Thanks @henryiii for your review.

> I think the wheels also need to be dual tagged. You can tell maturin to do this. `--compatibility 2_17` IIRC.

Can you open a new PR for that? I think it's out of scope of this PR.

**Comment by charliermarsh** [comment]
I’m undecided on whether I want to do this. But we should either (1) do this and omit all the non-musl release targets (this would now be considered a breaking change), (2) do this but copy the musl to the non-musl targets (sort of strange IMO), or (3) re-add the GNU dynamically linked build (but keep the dual-tagged wheel).

**Comment by charliermarsh** [comment]
Are there examples of other similar tools that only ship against musl?

**Comment by ofek** [comment]
ripgrep does that but only for x86_64 https://github.com/BurntSushi/ripgrep/releases/tag/14.1.0

**Comment by BurntSushi** [comment]
For ripgrep, I don't really have a consistent philosophy here. I'm pretty reactive and add things that people request (or submit), and will generally remove artifacts if they become too burdensome to maintain. With that said, if I were to clean things up, I'd probably bias toward using musl for everything where I could, and not offer anything that is dynamically linked to glibc. I don't think I've ever offered a glibc release artifact for Linux x86-64, and I don't think I've ever heard anyone complain about it. Using musl is just more convenient and doesn't really confer any downsides for ripgrep specifically. (Its allocator tends to be slower than glibc, but I mitigate that by using a different allocator. I think we do the same for `uv`.)

**Comment by charliermarsh** [comment]
Ok, I think for now, my preference is:

- Keep these builds.
- Re-add the dynamically-linked GNU variant to the release pipeline.
- ...but dual-tag our wheels, and upload the dual-tagged wheels.

**Comment by T-256** [comment]
Also see https://www.tweag.io/blog/2023-08-10-rust-static-link-with-mimalloc/

> (Its allocator tends to be slower than glibc, but I mitigate that by using a different allocator. I think we do the same for `uv`.)

Correct, according to blog post with mimalloc it'd be even faster than linux-gnu target. We currently use jemalloc.

**Comment by charliermarsh** [comment]
Yeah I'm ultimately a bit torn but it doesn't really cost us much to continue shipping the dynamically-linked versions, so I have a slight preference towards doing so.

**Comment by T-256** [comment]
According to https://github.com/astral-sh/uv/pull/4254, this is now breaking change, 
@charliermarsh feel free to close it since you agreed continuation with both static and dynamic libc releases in the pipeline.

**Pull request closed without merging**